src/Lib.P2P.Tests/Transports/TcpTest.cs

```diff
@@ -0,0 +1,224 @@
+            var server = await new MultiAddress("/dns4/time.nist.gov/tcp/37").ResolveAsync(cs.Token);
+            var data = new byte[4];
+
+            var tcp = new Tcp();
+            using (var time = await tcp.ConnectAsync(server[0], cs.Token))
+            {
+                var n = await time.ReadAsync(data, 0, data.Length, cs.Token);
+                Assert.AreEqual(4, n); // sometimes zero!
+            }
+        }
+
+        [TestMethod]
+        public void Listen_Then_Cancel()
+        {
+            var tcp = new Tcp();
+            var cs = new CancellationTokenSource();
+            MultiAddress listenerAddress = null;
+            Action<Stream, MultiAddress, MultiAddress> handler = (stream, local, remote) =>
+            {
+                Assert.Fail("handler should not be called");
+            };
+            listenerAddress = tcp.Listen("/ip4/127.0.0.1", handler, cs.Token);
+            Assert.IsTrue(listenerAddress.Protocols.Any(p => p.Name == "tcp"));
+            cs.Cancel();
+        }
+
+        [TestMethod]
+        public async Task Listen()
+        {
+            var tcp = new Tcp();
+            var cs = new CancellationTokenSource(TimeSpan.FromSeconds(30));
+            var connected = false;
+            MultiAddress listenerAddress = null;
+            Action<Stream, MultiAddress, MultiAddress> handler = (stream, local, remote) =>
+            {
+                Assert.IsNotNull(stream);
+                Assert.AreEqual(listenerAddress, local);
+                Assert.IsNotNull(remote);
+                Assert.AreNotEqual(local, remote);
+                connected = true;
+            };
+            try
+            {
+                listenerAddress = tcp.Listen("/ip4/127.0.0.1", handler, cs.Token);
+                Assert.IsTrue(listenerAddress.Protocols.Any(p => p.Name == "tcp"));
+                using (var stream = await tcp.ConnectAsync(listenerAddress, cs.Token))
+                {
+                    await Task.Delay(50);
+                    Assert.IsNotNull(stream);
+                    Assert.IsTrue(connected);
+                }
+            }
+            finally
+            {
+                cs.Cancel();
+            }
+        }
+
+        [TestMethod]
+        public async Task Listen_Handler_Throws()
+        {
+            var tcp = new Tcp();
+            var cs = new CancellationTokenSource(TimeSpan.FromSeconds(30));
+            var called = false;
+            Action<Stream, MultiAddress, MultiAddress> handler = (stream, local, remote) =>
+            {
+                called = true;
+                throw new Exception("foobar");
+            };
+            try
+            {
+                var addr = tcp.Listen("/ip4/127.0.0.1", handler, cs.Token);
+                Assert.IsTrue(addr.Protocols.Any(p => p.Name == "tcp"));
+                using (var stream = await tcp.ConnectAsync(addr, cs.Token))
+                {
+                    await Task.Delay(50);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)